resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1450,10 +1422,78 @@ final Future<AddressedEnvelope<DnsResponse, InetSocketAddress>> doQuery(
+            Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> inflight =
+                    inflightLookups.get(question);
+            if (inflight != null) {
+                inflight.addListener(f -> {
+                    if (f.isSuccess()) {
+                        // Notify the observer and after that the promise
+                        queryLifecycleObserver.querySucceed();
+                        AddressedEnvelope<? extends DnsResponse, InetSocketAddress> result =
+                                (AddressedEnvelope<? extends DnsResponse, InetSocketAddress>) f.getNow();
+
+                        // Retain the result as the listener on the promise is responsible to release it.
+                        ReferenceCountUtil.retain(result);
+                        promise.setSuccess(result);
+                    } else {
+                        Throwable cause = f.cause();
+                        if (isTimeoutError(cause)) {
+                            doQueryNow(channel, nameServerAddr, question, queryLifecycleObserver,
+                                    additionals, flush, payloadSize, castPromise);
+                        } else {
+                            // Notify the observer and after that the promise
+                            queryLifecycleObserver.queryFailed(cause);
+                            promise.setFailure(cause);
+                        }
+                    }
+                });
+            } else if (inflightLookups.size() < maxNumConsolidation) {
+                // Create a new promise as we need to ensure we are the first that will add a listener to it to retain
+                // the result before anyone can release it.
+                Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> newPromise =
+                        executor().newPromise();
+                Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> old =
+                        inflightLookups.put(question, newPromise);
+                assert old == null;
+
+                newPromise.addListener(f -> {
+                    if (f.isSuccess()) {
+                        // On success we need to retain the result so listeners that are added after this one
+                        // will still be able to use the result.
+                        AddressedEnvelope<? extends DnsResponse, InetSocketAddress> result =
+                                (AddressedEnvelope<? extends DnsResponse, InetSocketAddress>) f.getNow();
+                        ReferenceCountUtil.retain(result);
+                    }
+                    // Remove the promise and add another listener to it that will call release() on the result.
+                    // As the execution of the listeners is guaranteed to be in the same order as how these were added
+                    // we know that all previous added listeners had a chance to handle the result already.
```

Comment:
Is this true even if the promise is in the process of executing its listeners? I _believe_ `f == newPromise == inflightLookups.get(question)`.